### PR TITLE
fix: update with changes for OpenSSL 3.x

### PR DESCRIPTION
**Pull request by danbev**
This commit updates __init__.py with changes that were made directly
in Node.js (sorry about this but at the time I was not aware that there
this upstream repo existed).

It includes changes from two commits in Node.js:
 Commit f4bd91b0e2d7f07365ccf6bba943ba97d6f95cc9 ("deps,build,
tools: fix openssl-is-fips for ninja builds) and
Commit 66da32c045035cf2710a48773dc6f55f00e20c40 ("deps,test,src,doc,
tools: update to OpenSSL 3.0")

Refs: https://github.com/nodejs/node/issues/40735

### CODE REVIEW

**Comment by danbev** [comment]
This change is also available in https://github.com/nodejs/gyp-next/pull/147 but I was not sure if you preferred to have a separate commit. Also sorry about not making the changes upstream first, it was just lack of knowledge on my part about how project is used by Node.js.

File: pylib/gyp/__init__.py [hunk 1]
```diff
@@ -103,6 +103,14 @@ def Load(
     for (key, val) in generator.generator_default_variables.items():
         default_variables.setdefault(key, val)
 
+    output_dir = params["options"].generator_output or params["options"].toplevel_dir
+    if (default_variables['GENERATOR'] == 'ninja'):
```
Comment (cclauss) [inline pylib/gyp/__init__.py:107]:
```suggestion
    if default_variables['GENERATOR'] == 'ninja':
```

Comment (danbev) [reply pylib/gyp/__init__.py]: I've updated this now, thanks!

**Review by ryzokuken — APPROVED** [review]

**Comment by gengjiawen** [comment]
Is changes `pylib/gyp/generator/make.py` no longer needed ?

File: pylib/gyp/__init__.py [hunk 2]
```diff
@@ -103,6 +103,14 @@ def Load(
...
         default_variables.setdefault(key, val)
 
+    output_dir = params["options"].generator_output or params["options"].toplevel_dir
+    if default_variables['GENERATOR'] == 'ninja':
+        default_variables.setdefault("PRODUCT_DIR_ABS", os.path.join(output_dir,
+                                     'out', default_variables['build_type']))
+    else:
+        default_variables.setdefault("PRODUCT_DIR_ABS", os.path.join(output_dir,
+                                     default_variables['build_type']))
+
```
Comment (cclauss) [inline pylib/gyp/__init__.py:113]: OPTIONAL: Readability...
```suggestion
    if default_variables['GENERATOR'] == 'ninja':
        prod_dir = os.path.join(output_dir, 'out', default_variables['build_type'])
    else:
        prod_dir = os.path.join(output_dir, default_variables['build_type'])
    default_variables.setdefault("PRODUCT_DIR_ABS", prod_dir)
```

Comment (danbev) [reply pylib/gyp/__init__.py]: I've made these changes now, thanks.

**Comment by danbev** [comment]
> Is changes pylib/gyp/generator/make.py no longer needed ?

I'll take a closer look but just by looking at them I really don't recognize those changes, but then again it was a long time since they were made.

**Review by cclauss — APPROVED** [review]

**Comment by gengjiawen** [comment]
> > Is changes pylib/gyp/generator/make.py no longer needed ?
> 
> I'll take a closer look but just by looking at them I really don't recognize those changes, but then again it was a long time since they were made.

comes from https://github.com/nodejs/node/pull/38512

**Comment by danbev** [comment]
> comes from [nodejs/node#38512](https://github.com/nodejs/node/pull/38512)

Thanks, I saw that it was included by that commit, but what I can't remember is the reason for these changes. I have a vague memory that this was a change during early development where there were some command line options that became too long (something like that). I've tried reverting the changes to make.py and running the Node.js CI tests to see if anything breaks: https://github.com/nodejs/node/pull/43051

**Comment by targos** [comment]
To make the changelog entry clearer about the change, what commit message should we use?

I'd start with this:

```
feat: add `PRODUCT_DIR_ABS` variable
```

**Comment by targos** [comment]
It would be nice to explain what directory the variable points to and what it can be used for.

**Comment by targos** [comment]
Why close this? I thought it made sense to separate the `__init__` and `make` changes to better document them in the changelog.

**Comment by danbev** [comment]
> Why close this? I thought it made sense to separate the `__init__` and `make` changes to better document them in the changelog.

My reason for opening this PR was I did not think all the changed in the other PR were required, but I was wrong about that. If you want to split #147 in multiple commits then that sounds good to me.

**Pull request closed without merging**